### PR TITLE
Revert "Improve the code gen for a conditional access on a readonly unconstrained field"

### LINKED ISSUE #40690: System.AccessViolationException when invoking extension method on ValueTuple
# Issue Title

With .NET Core 3.1.100 we're getting a:
`Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.` 
when invoking an extension method on a field of a `ValueTuple`.

This was fine with .NET Core 3.0.100

# General

We're trying to update the Quantum Development Kit runtime from `3.0.100` to `3.1.100` and one of our tests is failing and crashing execution with the exception mentioned above. 

I have pinpoint the problem, it happens when we're trying to call an extension method (an extension for `object`) of a field of a ValueTuple if the field is a generic Type Parameter, for example:

```C#
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace bug
{
    public static class Extension
    {
        public static String GetValue(this object value) => value?.ToString();
    }

    public partial class BGen<__T__>
    {
        public (Int64, __T__) Data { get; }

        public BGen((Int64, __T__) data)
        {
            this.Data = data;
        }

        internal String Value
        {
            get
            {
                /// This works fine:
                //var i2 = Data.Item2;
                //var q2 = i2?.GetValue();

                // This causes the AccessViolation:
                var q2 = Data.Item2?.GetValue();
                return q2;
            }
        }
    }
    
    class Program
    {
        static void Main(string[] args)
        {
            var i = new BGen<string>((0, "foo"));
            var r = i.Value;

            Console.WriteLine($"Hello World: {string.Join(',', r)}");
        }
    }
}
```

This happens on windows with version `3.1.100`. Works fine if downgrading to `3.0.100`.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Reverts #35979
Closes #40690 

We found that this change makes the compiler generate invalid IL in some cases. For now we have decided that the right thing to do is to back the change out while keeping the tests (adjusting for the IL now produced after reverting).

@YairHalberstadt sorry that we have to revert this. We would be happy to consider a future PR that restores this optimization while ensuring that more corner cases like the one in the linked issue are handled.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConditionalOperatorTests.cs [hunk 1]
```diff
@@ -3025,25 +3034,118 @@ public static void Main()
...
+  IL_002b:  br.s       IL_0038
+  IL_002d:  constrained. ""T""
+  IL_0033:  callvirt   ""string object.ToString()""
+  IL_0038:  call       ""void System.Console.WriteLine(string)""
+  IL_003d:  ret
 }");
+        }
+
+        [Fact, WorkItem(40690, "https://github.com/dotnet/roslyn/issues/40690")]
+        public void ConditionalAccess_GenericExtension_ValueTuple()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConditionalOperatorTests.cs]: >ConditionalAccess_GenericExtension_ValueTuple [](start = 20, length = 45)

Is there an equivalent test for VB that fails without the change?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenConditionalOperatorTests.cs]: I think this test case would translate pretty straightforwardly into VB. I can add a regression test for it.

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 2]
```diff
@@ -405,43 +405,32 @@ private void EmitLoweredConditionalAccessExpression(BoundLoweredConditionalAcces
...
+                    EmitDefaultValue(receiverType, true, receiver.Syntax);
+                    EmitBox(receiverType, receiver.Syntax);
+                    _builder.EmitBranch(ILOpCode.Brtrue, whenNotNullLabel);
+                    EmitLoadIndirect(receiverType, receiver.Syntax);
+
+                    cloneTemp = AllocateTemp(receiverType, receiver.Syntax);
+                    _builder.EmitLocalStore(cloneTemp);
+                    _builder.EmitLocalAddress(cloneTemp);
+                    _builder.EmitLocalLoad(cloneTemp);
+                    EmitBox(receiver.Type, receiver.Syntax);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: >receiver.Type [](start = 28, length = 13)

Did this need to change? Same comment for VB.

**Comment by YairHalberstadt** [comment]
Oops!

I think I have an easy fix for this - I will try to push by the end of the day.

It would be useful if you could hold off from merging to avoid merge conflicts and messing up the git history?

Thanks

**Comment by YairHalberstadt** [comment]
Fix at https://github.com/dotnet/roslyn/pull/40878

**Comment by RikkiGibson** [comment]
We decided that it will be better to take the revert for the short term (16.5 release) in order to be able to provide a fix that we have higher confidence in before that release window closes.

We are still interested in this optimization, potentially for [16.6](https://github.com/dotnet/roslyn/tree/release/dev16.6-preview1), when we have a little more flexibility, more time to reason through the scenarios and gain confidence that the codegen for this optimization is completely solid.

Thanks for understanding @YairHalberstadt and thank you for your prompt response when the issue came up.

**Pull request merged**